### ISSUE TITLE
Cannot bind appserver to port 80 on macOS, works on Windows

### ISSUE THREAD

**Issue by ptmkenny** [lando/lando#352]
Bug Report
==========

I can't get the appserver port to bind to port 80 on macOS, but it works on Windows.

I'm using a Pantheon recipe with beta17.  I have installed lando on both a Windows PC and a Mac.

The Windows machine uses port 80, but the Mac randomly picks a port like 3xxxx. When I use `lsof -i:80` to check what is on port 80, the result is `vpnkit`, which is part of docker:

```
COMMAND  PID USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
vpnkit  4164   pk   29u  IPv4 0xfa3bbef8a0e19cdd      0t0  TCP 192.168.0.106:63175->mirror-isc3.debian.org:http (CLOSE_WAIT)
vpnkit  4164   pk   37u  IPv4 0xfa3bbef8afa85cdd      0t0  TCP 192.168.0.106:63176->dcsx12.anu.edu.au:http (CLOSE_WAIT)
vpnkit  4164   pk   43u  IPv4 0xfa3bbef8b3313cdd      0t0  TCP 192.168.0.106:62872->dcsx12.anu.edu.au:http (CLOSE_WAIT)
vpnkit  4164   pk   49u  IPv4 0xfa3bbef89ccb81f5      0t0  TCP 192.168.0.106:64227->dcsx12.anu.edu.au:http (CLOSE_WAIT)
vpnkit  4164   pk   50u  IPv4 0xfa3bbef89d2641f5      0t0  TCP 192.168.0.106:63599->senfter.debian.org:http (CLOSE_WAIT)
vpnkit  4164   pk   63u  IPv4 0xfa3bbef8b32ca1f5      0t0  TCP 192.168.0.106:64200->mirror-csail.debian.org:http (CLOSE_WAIT)
```

I'm using the same  `.lando.yml` on both the Mac and Windows.

```
# If you change this file, lando rebuild.

# Change php and docroot in .pantheon.yml, not here.

name: mysite
recipe: pantheon
config:
  framework: drupal8
  xdebug: true
  env: dev
  site: mysite
  id: 2d533f7c-4fcb-47fd-b446-fe89c364f57b
  webroot: web/

# Additional mailhog proxy settings so we get a nice URL for the mailhog ui
proxy:
  mailhog:
    - mail.mysite.lndo.site

services:
  database:
    portforward: 3306
  appserver:
    # portforward: 80
    extras:
      # !! Commands run as ROOT.  Don't fuck it up.
      - "apt-get update -y"
      - "apt-get install vim -y"
    overrides:
      services:
        # ports: ['80:80']
        environment:
          BEHAT_PARAMS: '{"extensions" : {"Behat\\MinkExtension" : {"base_url" : "http://mysite.lndo.site/"}, "Drupal\\DrupalExtension" : {"drush" :   {  "root":  "/app/web" }}}}'

tooling:
  behat:
    service: appserver
    description: Run behat tests locally.
    cmd:
      - /app/vendor/bin/behat
```

macOS 10.12.6 using beta17

`docker ps`

```
CONTAINER ID        IMAGE                             COMMAND                  CREATED             STATUS              PORTS                                                                 NAMES
44abd9811539        nginx:latest                      "/lando-entrypoint..."   9 minutes ago       Up 9 minutes        80/tcp, 0.0.0.0:32822->443/tcp                                        mysite_edge_ssl_1
d83e3b2924c1        eeacms/varnish:4.1-3.0            "/lando-entrypoint..."   9 minutes ago       Up 9 minutes        6081/tcp, 0.0.0.0:32821->80/tcp                                       mysite_edge_1
cdf0b74f8de6        nginx:1.8                         "/lando-entrypoint..."   9 minutes ago       Up 9 minutes        0.0.0.0:32819->80/tcp, 0.0.0.0:32818->443/tcp                         mysite_nginx_1
543a3de33a48        mailhog/mailhog:latest            "/lando-entrypoint..."   9 minutes ago       Up 9 minutes        8025/tcp, 0.0.0.0:32817->80/tcp, 0.0.0.0:1026->1025/tcp               mysite_mailhog_1
aa7388658b8a        mariadb:10.0                      "/lando-entrypoint..."   9 minutes ago       Up 9 minutes        0.0.0.0:3306->3306/tcp                                                mysite_database_1
b488ec2fae66        devwithlando/pantheon-index:3.6   "/lando-entrypoint..."   9 minutes ago       Up 9 minutes        0.0.0.0:32814->449/tcp                                                mysite_index_1
501d8e7f501b        redis:2.8                         "/lando-entrypoint..."   9 minutes ago       Up 9 minutes        0.0.0.0:32815->6379/tcp                                               mysite_cache_1
d299cca0fbdc        traefik:1.3-alpine                "/lando-entrypoint..."   2 hours ago         Up 13 minutes       0.0.0.0:443->443/tcp, 0.0.0.0:8000->80/tcp, 0.0.0.0:58086->8080/tcp   landoproxyhyperion5000gandalfedition_proxy_1

```

Windows 10 Pro beta 17

`docker ps`

```
CONTAINER ID        IMAGE                                     COMMAND                  CREATED             STATUS              PORTS
               NAMES
0a1b95a1e8ff        nginx:latest                              "/lando-entrypoint..."   2 minutes ago       Up 2 minutes        80/tcp, 0.0.0.0:32825->443/tcp
               mysite_edge_ssl_1
4d05b893b8a8        eeacms/varnish:4.1-3.0                    "/lando-entrypoint..."   2 minutes ago       Up 2 minutes        6081/tcp, 0.0.0.0:32824->80/tcp
               mysite_edge_1
7335d2c88389        nginx:1.8                                 "/lando-entrypoint..."   2 minutes ago       Up 2 minutes        0.0.0.0:32822->80/tcp, 0.0.0.0:32821->443/tcp
               mysite_nginx_1
30b4550500c1        devwithlando/pantheon-appserver:7.0-fpm   "/lando-entrypoint..."   2 minutes ago       Up 2 minutes        9000/tcp
               mysite_appserver_1
009e1db03a48        redis:2.8                                 "/lando-entrypoint..."   9 minutes ago       Up 2 minutes        0.0.0.0:32818->6379/tcp
               mysite_cache_1
5d84b532d992        mailhog/mailhog:latest                    "/lando-entrypoint..."   9 minutes ago       Up 2 minutes        8025/tcp, 0.0.0.0:32820->80/tcp, 0.0.0.0:1026->1025/tcp             mysite_mailhog_1
1718ef2c464c        mariadb:10.0                              "/lando-entrypoint..."   9 minutes ago       Up 2 minutes        0.0.0.0:32817->3306/tcp
               mysite_database_1
e99cfe292dc5        devwithlando/pantheon-index:3.6           "/lando-entrypoint..."   9 minutes ago       Up 2 minutes        0.0.0.0:32819->449/tcp
               mysite_index_1
63853c85335c        traefik:1.3-alpine                        "/lando-entrypoint..."   7 hours ago         Up 2 minutes        0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp, 0.0.0.0:58086->8080/tcp   landoproxyhyperion5000gandalfedition_proxy_1

```

I tried using `ports: ['80:80']` as suggested in the documentation. I ran `lando rebuild` but I got this error:

```
ERROR: for 4ded8a46922e_4ded8a46922e_mysite_appserver_1  Cannot start service appserver: driver failed programming external connectivity on endpoint mysite_appserver_1 (09587487a65783e9b27a44a71c1a0d2cd75cbe6fedbb93c499f7d430028eebf7): Bind for 0.0.0.0:32819 failed: port is already allocated

ERROR: for appserver  Cannot start service appserver: driver failed programming external connectivity on endpoint mysite_appserver_1 (09587487a65783e9b27a44a71c1a0d2cd75cbe6fedbb93c499f7d430028eebf7): Bind for 0.0.0.0:32819 failed: port is already allocated
ERROR: Encountered errors while bringing up the project.
```

Other things I tried that did not work:
* Reset docker to factory defaults
* Tried using "No proxy" in docker settings (instead of "Use system proxy"
* Turned off the macOS firewall

Other notes: I have kalabox installed on this machine if that matters.

**Comment by jbertoen** [comment]
It has probably something to do with the proxy container. Do you have apache running on OSX, it was already installed on my OSX environment so I had to kill it and disable it to get it working on port 80. You can check this by doing "ps aux | grep http" in your terminal and kill the process doing a "sudo kill [pid]"

d299cca0fbdc        traefik:1.3-alpine                "/lando-entrypoint..."   2 hours ago         Up 13 minutes       0.0.0.0:443->443/tcp, 0.0.0.0:8000->80/tcp, 0.0.0.0:58086->8080/tcp   landoproxyhyperion5000gandalfedition_proxy_1

**Comment by pirog** [comment]
@ptmkenny if you are going to set your own static ports then its up to you to set them up correctly. For that reason this is not really a problem with lando itself but with your configuration. That said and generally speaking its probably not a great idea to statically bind a single app or service to port 80 (or any port for that matter) because the chance of collision with something else is pretty high.

As @jbertoen points out Lando does use a proxy (which is bound to port 80 by default if available) to help mitigate some of these collisions and its definitely recommended to just use that. Is there a reason why you are not? https://docs.lndo.io/config/proxy.html

**Comment by ptmkenny** [comment]
@pirog I am just trying to use lando on port 80 on Mac, but when I start it, it uses port 8000.  (I want it on the same port on Mac and Windows because I have a bunch of bookmarks that open in Chrome to various admin pages, and if Lando is on different ports, all of these break.)

As @jbertoen said I did have apache running, which I shut off with `sudo apachectrl stop`.  Still, after restarting docker and lando, I still can't get it to go to port 80.  But I understand that's not a lando problem so I'll look elsewhere to figure that out.

That said, on a lark I tried fixing the port on Windows as specified in the docs, but no matter what port I tried, I got the error

```
    overrides:
      services:
        ports: ['1234:80']
```

```
ERROR: for appserver  Cannot start service appserver: driver failed programming external connectivity on endpoint mysite_appserver_1 (09587487a65783e9b27a44a71c1a0d2cd75cbe6fedbb93c499f7d430028eebf7): Bind for 0.0.0.0:1234 failed: port is already allocated
ERROR: Encountered errors while bringing up the project.

```
